src/openrct2/ride/TrackDesign.cpp

```diff
@@ -187,6 +187,16 @@ rct_string_id TrackDesign::CreateTrackDesignTrack(const Ride& ride)
     auto trackType = trackElement.element->AsTrack()->GetTrackType();
     uint8_t direction = trackElement.element->GetDirection();
     _saveDirection = direction;
+
+    // ride_get_start_of_track returns the x and y of the first section of track, but the z location
+    // of the last section so we correct it by adding the difference
+    const rct_preview_track* trackSection = get_track_def_from_ride(&ride, trackType);
+    int32_t firstSectionZ = trackSection->z;
+    trackSection += trackElement.element->AsTrack()->GetSequenceIndex();
```

Comment:
I'd like @duncanspumpkin 's opinion on this block. I really can't say if it is correct or not.

Reply:
It is correct but i want to do a few tests of this pr before merging

Reply:
If it helps, here's a save file where I've spammed rollercoasters with different starting pieces to test them: [PR12676.zip](https://github.com/OpenRCT2/OpenRCT2/files/5077572/PR12676.zip)

Reply:
I think I've figured out exactly why it's broken. It's to do with how `track_block_get_previous` / `track_block_get_previous_from_zero` work. They use a struct to store their result that contains the 5 relevent fields: begin_x, begin_y, end_x, end_y, begin_element. However, because `track_block_get_previous` traverses backwards, the begin_element label is misleading: when the algorithm looks for the previous track piece, it looks for a track section in location begin_x and begin_y. The piece it finds is the last section of the previous piece of track, not the first. `ride_get_start_of_track` would then return the coordinates for the first track section, but the element for the last track section. See the below picture if that's confusing...

![this picture](https://user-images.githubusercontent.com/2857218/90302284-04cd0c00-de9d-11ea-84aa-e9a7f802b923.png)

I've completely reworked this PR to just fix `ride_get_start_of_track` to return the coordinates for the last section instead of the coordinates for the first section (begin_element along with end_x and end_y). Both places where this function is used, only the first track piece is needed, not the first section of track, so this isn't a breaking change (the other place is for checking the refund amount for rides, which I also tested).